### PR TITLE
New optimization: DimShuffle(Alloc(...)) -> Alloc(..., use dims of 1 for new dimensions) #5024

### PR DESCRIPTION
**Pull request by olimastro**
PR for new optimization dealing with 

`dimshuffle{x, 0, 1}(alloc([3 4], 3, 2) => alloc([3 4], 1, 3, 2)` 

in uncanonicalize phase

gh-5024

### CODE REVIEW

File: theano/tensor/opt_uncanonicalize.py [hunk 1]
```diff
@@ -146,3 +146,29 @@ def local_reshape_dimshuffle(node):
...
+    """
+    If an alloc is inside a dimshuffle which only adds dimension to the left,
+    scrap the dimshuffle and adds 1 into the alloc
+
+    dimshuffle{x, 0, 1}(alloc([3 4], 3, 2) => alloc([3 4], 1, 3, 2)
+    """
+    if isinstance(node.op, T.DimShuffle):
+        #import ipdb ; ipdb.set_trace()
+        # check if it only adds dimension to the left
+        new_order = node.op.new_order
```
Comment (nouiz) [inline theano/tensor/opt_uncanonicalize.py]: you need to check that node.inputs[0].owner.op is an Alloc.

File: theano/tensor/tests/test_opt_uncanonicalize.py [hunk 2]
```diff
@@ -145,3 +146,25 @@ def test_local_reshape_dimshuffle():
...
     topo = g.toposort()
     assert any([not isinstance(x, DimShuffle) for x in topo])
+
+
+
+def test_local_reshape_dimshuffle():
+
+    reshape_dimshuffle = out2in(local_dimshuffle_alloc)
+
+    x = tensor.matrix('x')
```
Comment (nouiz) [inline theano/tensor/tests/test_opt_uncanonicalize.py]: Make this a vector. On line 167, you give a vector as input.

File: theano/tensor/tests/test_opt_uncanonicalize.py [hunk 3]
```diff
@@ -145,3 +146,25 @@ def test_local_reshape_dimshuffle():
...
+
+    out = tensor.alloc(x, 3, 2).dimshuffle('x','x',0,1)
```
Comment (nouiz) [inline theano/tensor/tests/test_opt_uncanonicalize.py]: pep8, 1 space after comma.

**Comment by olimastro** [comment]
You are right I went too fast!

**Comment by nouiz** [comment]
There is still this error:

  File "/home/travis/build/Theano/Theano/theano/tensor/opt_uncanonicalize.py", line 160, in local_dimshuffle_alloc
    if isinstance(node.op, T.DimShuffle) and isinstance(node.inputs[0].owner.op, T.Alloc):
AttributeError: 'NoneType' object has no attribute 'op'

File: theano/tensor/opt_uncanonicalize.py [hunk 4]
```diff
@@ -157,7 +157,7 @@ def local_dimshuffle_alloc(node):
 
     dimshuffle{x, 0, 1}(alloc([3 4], 3, 2) => alloc([3 4], 1, 3, 2)
     """
-    if isinstance(node.op, T.DimShuffle) and isinstance(node.inputs[0], T.Alloc):
+    if isinstance(node.op, T.DimShuffle) and isinstance(node.inputs[0].owner.op, T.Alloc):
```
Comment (nouiz) [inline theano/tensor/opt_uncanonicalize.py]: node.inputs[0].owner can be done.  This is the cause of the error in travis.

**Comment by nouiz** [comment]
jenkins test this

**Pull request merged**